### DISCUSSION TITLE
What is the difference between Restart and Redeploy?

### DISCUSSION THREAD

**Discussion by predragnikolic** [coollabsio/coolify#2935] [Q&A]
Hello 👋 

What is the difference between the Restart and Redeploy buttons?
It looks like they do the same thing currently and that is to start a deployment.

When I first saw the restart button, I tought  it stops the app process, and start it again. But that is not what the button actually do! It looks like it just triggers a deployment. And that is what the "redeploy" button also does.

So what is the difference and when should I use which button?
<img width="585" alt="Screenshot 2024-07-24 at 17 01 48" src="https://github.com/user-attachments/assets/29bb6f3e-0105-4a35-8bcb-1e84207a1bd8">

**Comment by franciscohanna92** [comment]
Confused by this also!

**Comment by henningsieh** [comment]
Wow, no answer in 6 month? Cool...

**Comment by oliverfrost1** [comment]
It would be great with a simple explainer tooltip over them. A thing I would like clarified is if a redeploy closes the server while the redeployment happens. Also, if the redeployment fails, will it keep my current instance running?

**Comment by srinisubramanian** [comment]
As far as I can make out a restart would just restart the container.  A redeploy probably does a rebuild?

**Comment by nktnet1** [comment]
Yeah, not entirely clear in the docs, although there are [hints in the source code](https://github.com/coollabsio/coolify/blob/6062fa7b4d06afdbeb9f7723cef95fe8eec1b53d/resources/views/livewire/project/application/heading.blade.php#L36-L78).

Behaviour for swarm mode is also different, but for your regular deployment:

![html-code](https://github.com/user-attachments/assets/da5650a4-8a2a-4ab1-84c3-f06d50639218)

It appears that "redeploy" will rebuild the docker image, whereas "restart" won't.

We should add an entry to the FAQ (will do when I find time, but all welcome):
- https://coolify.io/docs/knowledge-base/faq
- https://github.com/coollabsio/documentation-coolify/tree/main/docs/knowledge-base/faq.md

**Reply by smeijer** [reply]
> It appears that "redeploy" will rebuild the docker image, whereas "restart" won't.

That's not the behaviour I'm seeing. I still see docker rebuilds (using a build server) when simply pressing "restart".

**Reply by nktnet1** [reply]
Did you do either of the following?

1. Change configurations ([potentially environment variables](https://github.com/coollabsio/coolify/discussions/2935#discussioncomment-13508923))
1. Prune docker (e.g. delete cache/images/volume/etc) - there is also automatic docker clean-up to look out for

If restarting without either of the above still results in a rebuild, you may want to report it as an issue with reproduction steps ^.

**Comment by arthureberledev** [comment]
Do you guys know if a redeploy will deploy the latest changes from a git repo? I have disabled the automatic builds on updates to the repo.

**Reply by nktnet1** [reply]
Yes - the latest commit in the branch you've set.

**Comment by ahmafi** [comment]
Restart is triggering a new deploy! I just change some ENV variables. Is this supposed to happen?

**Reply by smeijer** [reply]
Exactly what i just experienced. It's confusing, and not helpful (giving the long build time of this container)

**Reply by nktnet1** [reply]
Speculating, but possibly the problem here is that sometimes environment variables are used during the build stage of Docker (i.e. not just at run time). Since "restarting" will not apply such changes, Coolify opts for rebuilding the image altogether.

**Comment by feschaffa** [comment]
Also confused by this, restart is rebuilding the whole image 🤔

**Comment by djsisson** [comment]
restart will be set to false if its a docker image or dockerfile (no idea why) (also doesn't show if compose)

https://github.com/coollabsio/coolify/blob/723f14e6f6a84bab744a792c3e628c5e5a3fa308/app/Jobs/ApplicationDeploymentJob.php#L197

and same if anything in the configuration has changed, but that should be a given

https://github.com/coollabsio/coolify/blob/723f14e6f6a84bab744a792c3e628c5e5a3fa308/app/Jobs/ApplicationDeploymentJob.php#L846
